core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericSummaryParallelIterator.java

```diff
@@ -0,0 +1,355 @@
+    
+    // TODO
+    expect_sums_and_counts = false;
+    
+    DownsampleConfig downsampleConfig = ((GroupBy) node).getDownsampleConfig();
+    if (null == downsampleConfig) {
+      throw new IllegalStateException("Shouldn't be here if the downsample was null.");
+    }
+    intervals = downsampleConfig.intervals();
+    interval = downsampleConfig.interval();
+    interval_in_seconds = DateTime.parseDuration(downsampleConfig.getInterval()) / 1000;
+    start_epoch = (int) downsampleConfig.startTime().epoch();
+    ts = downsampleConfig.startTime().getCopy();
+    System.out.println("        START: " + ts.epoch());
+    agg = AggEnum.valueOf(((GroupByConfig) node.config()).getAggregator().toLowerCase());
+    infectious_nan = ((GroupByConfig) node.config()).getInfectiousNan();
+    dp = new MutableNumericSummaryValue();
+    accumulators = new Accumulator[GroupByNumericArrayIterator.NUM_THREADS];
+    for (int i = 0; i < accumulators.length; i++) {
+      accumulators[i] = new Accumulator(i);
+    }
+
+    List<Future<Void>> futures = 
+        new ArrayList<>(sources.size());
+    int i = 0;
+    for (final TimeSeries source : sources) {
+      int index = (i++) % GroupByNumericArrayIterator.NUM_THREADS;
+      final ExecutorService executorService = 
+          GroupByNumericArrayIterator.executorList.get(index);
+      
+      final long s = System.nanoTime();
+      Future<Void> future =
+          executorService.submit(
+              () -> {
+                //statsCollector.addTime("groupby.queue.wait.time", System.nanoTime() - s, ChronoUnit.NANOS);
+                accumulators[index].accumulate(source);
+                return null;
+              });
+
+      futures.add(future);
+    }
+    
+    for (Future<Void> future : futures) {
+      try {
+        future.get(); // get will block until the future is done
+      } catch (InterruptedException e) {
+        throw new QueryDownstreamException(e.getMessage(), e);
+      } catch (ExecutionException e) {
+        throw new QueryDownstreamException(e.getMessage(), e);
+      }
+    }
+    
+    combine();
+  }
+  
+  @Override
+  public boolean hasNext() {
+    return idx < results.length;
+  }
+
+  @Override
+  public TimeSeriesValue<? extends TimeSeriesDataType> next() {
+    dp.resetTimestamp(ts);
+    if (Double.isInfinite(results[idx])) {
+      dp.resetValue(expected_summary, Double.NaN);
+      idx++;
+    } else {
+      dp.resetValue(expected_summary, results[idx++]);
+    }
+    ts.add(interval);
+    return dp;
+  }
+
+  @Override
+  public TypeToken<? extends TimeSeriesDataType> getType() {
+    return NumericSummaryType.TYPE;
+  }
+  
+  void combine() {
+    results = new double[intervals];
+    Arrays.fill(results, Double.POSITIVE_INFINITY); // sentinel
+    
+    long[] counts = null;
+    if (agg == AggEnum.avg) {
+      counts = new long[intervals];
+    }
+    for (int i = 0; i < accumulators.length; i++) {
+      switch (agg) {
+      case sum:
```

Comment:
accumulate for sum aswell?

Reply:
Good catch

Reply:
oh wait, it was right the first time. this is for the sum OR count case. Count is handled in avg below if we want the avg.